test/forwarder/forwarder.js

```diff
@@ -105,61 +115,247 @@ contract('Forwarder', async (accounts) => {
+        const request_other =  {
+            from: relayer2Address,
+            to: TestTargetInstance.address,
+            value: '0',
+            gas: '300000',
+            nonce: 0,
+            data: '0x'
+        }
+
+        assert.equal((await ForwarderInstance.verify(request_other, sign)), false);
+        return TruffleAssert.reverts(ForwarderInstance.execute(request_other, sign), "MinimalForwarder: signature does not match request");
+    });
+
+    it ('If signature is valid, but req.nonce != nonce[signer], it should be reverted and should not be verified', async () => {
+        const request_other =  {
+            from: relayer1Address,
+            to: TestTargetInstance.address,
+            value: '0',
+            gas: '300000',
+            nonce: 10,
+            data: '0x'
+        }
+
+        assert.equal((await ForwarderInstance.verify(request_other, sign)), false);
+        return TruffleAssert.reverts(ForwarderInstance.execute(request_other, sign), "MinimalForwarder: signature does not match request");
+    });
+
+    it ('Execute should succeed even if the call to the target failed', async () => {
+        const new_domain = {
+            name,
+            version,
+            chainId: 1,
+            verifyingContract: ResponseForwarderInstance.address,
+        };
+
+        const new_request = {
+            from: relayer1Address,
+            to: ForwarderInstance.address,
+            value: '0',
+            gas: '300000',
+            nonce: 0,
+            data: '0x'
+        }
+
+        const new_sign = ethSigUtil.signTypedMessage(
+            relayer1.getPrivateKey(),
+            {
+                data: {
+                    types: types,
+                    domain: new_domain,
+                    primaryType: 'ForwardRequest',
+                    message: new_request
+                }
+            }
+        )
+
+        await ResponseForwarderInstance.response(new_request, new_sign);
+        const status = await ResponseForwarderInstance.status();
+        assert.equal(status, false);
+    });
+
+    it ('Should be failed in case of execute is called with less gas than req.gas', async () => {
+        const new_request = {
+            from: relayer3Address,
+            to: TestTargetInstance.address,
+            value: '0',
+            gas: '300000',
+            nonce: 0,
+            data: '0x'
+        }
+
+        const new_sign = ethSigUtil.signTypedMessage(
+            relayer3.getPrivateKey(),
+            {
+                data: {
+                    types: types,
+                    domain: domain,
+                    primaryType: 'ForwardRequest',
+                    message: new_request
+                }
+            }
+        )
+        const accountBalance = await provider.getBalance(accounts[0]);
+        await TruffleAssert.fails(ForwarderInstance.execute(new_request, new_sign, {value: accountBalance.toString()}));
```

Comment:
This one fails because the transaction itself is invalid due to insufficient balance.
To verify if the test is correct, try removing the `assert(gasleft() > req.gas / 63);` from Forwarder and see if test will fail.